idea/idea-fir/src/org/jetbrains/kotlin/idea/intentions/fir/ImportAllMembersIntention.kt

```diff
@@ -0,0 +1,113 @@
+        val classId = target.classIdIfNonLocal ?: return@inputProvider null
+        val selectionExpression = (psi.parent as? KtDotQualifiedExpression)?.selectorExpression
+        if (target.origin != KtSymbolOrigin.JAVA &&
+            (target.classKind == KtClassKind.OBJECT ||
+                    // One cannot use on-demand import for properties or functions declared inside objects
+                    isReferenceToObjectMemberOrUnresolved(selectionExpression))
+        ) {
+            // Import all members of an object is not supported by Kotlin.
+            return@inputProvider null
+        }
+        val shortenCommand = collectPossibleReferenceShortenings(
+            psi.containingKtFile,
+            classShortenOption = {
+                val queryClassId = it.classIdIfNonLocal ?: return@collectPossibleReferenceShortenings ShortenOption.DO_NOT_SHORTEN
+                if (queryClassId.packageFqName == classId.packageFqName && queryClassId.relativeClassName.parent() == classId.relativeClassName) {
+                    ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT
+                } else {
+                    ShortenOption.DO_NOT_SHORTEN
+                }
+            },
+            callableShortenOption = {
+                val containingClassId = if (it is KtConstructorSymbol) {
+                    it.containingClassIdIfNonLocal?.outerClassId
+                } else {
+                    it.callableIdIfNonLocal?.classId
+                }
+                if (containingClassId == classId) {
+                    ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT
+                } else {
+                    ShortenOption.DO_NOT_SHORTEN
+                }
+            }
+        )
+        if (shortenCommand.isEmpty) return@inputProvider null
+        Input(classId.asSingleFqName(), shortenCommand)
+    }
+
+    class Input(val fqName: FqName, val shortenCommand: ShortenCommand) : HLApplicatorInput
+
+    companion object {
+        val applicator = applicator<KtExpression, Input> {
+            familyName(KotlinBundle.lazyMessage("import.members.with"))
+            actionName { _, input -> KotlinBundle.message("import.members.from.0", input.fqName.asString()) }
+            isApplicableByPsi { it.isReceiverExpression() && !it.isInImportDirective() }
+            applyTo { _, input ->
+                input.shortenCommand.invokeShortening()
+            }
+        }
+
+        private fun KtExpression.isReceiverExpression(): Boolean {
+            return when (val parent = parent) {
+                is KtDotQualifiedExpression -> this == parent.receiverExpression
+                is KtUserType -> {
+                    val grandParent = parent.parent as? KtUserType ?: return false
+                    grandParent.qualifier == parent && parent.referenceExpression == this
+                }
+                else -> false
+            }
+        }
+
+        private fun KtExpression.reference(): KtReference? = when (this) {
```

Comment:
Ditto. This can be `val`.

Reply:
The tricky thing for this way is that a property called `referece` is already present... So if I want to make it a property, I need to come up with another name. Maybe `actualReference`?